src/app/GitUI/Editor/Diff/DiffHighlightService.cs

```diff
@@ -272,26 +284,66 @@ private static void MarkDifference(IDocument document, ISegment lineRemoved,
+            }
+        }
+        else
         {
-            color = AppColor.AnsiTerminalGreenBackBold.GetThemeColor();
-            markerStrategy.AddMarker(new TextMarker(lineAdded.Offset + beginOffset,
-                                                    lineAdded.Length - beginOffset - reverseOffset,
-                                                    TextMarkerType.SolidBlock, color,
-                                                    ColorHelper.GetForeColorForBackColor(color)));
+            int length = lineAdded.Length - lineStartOffset;
+            if (length > 0)
+            {
+                yield return CreatePaleMarker(lineAdded.Offset + lineStartOffset, length, GetAddedBackColor());
+            }
         }
 
-        if (lineRemoved.Length - beginOffset - reverseOffset > 0)
+        int removedLength = lineRemoved.Length - beginOffset - reverseOffset;
+        if (removedLength > 0)
         {
-            color = AppColor.AnsiTerminalRedBackBold.GetThemeColor();
-            markerStrategy.AddMarker(new TextMarker(lineRemoved.Offset + beginOffset,
-                                                    lineRemoved.Length - beginOffset - reverseOffset,
-                                                    TextMarkerType.SolidBlock, color,
-                                                    ColorHelper.GetForeColorForBackColor(color)));
+            int beforeLength = beginOffset - lineStartOffset;
+            if (beforeLength > 0)
+            {
+                yield return CreatePaleMarker(lineRemoved.Offset + lineStartOffset, beforeLength, GetRemovedBackColor());
+            }
+
+            int afterBegin = beginOffset + removedLength;
+            int afterLength = lineRemoved.Length - afterBegin;
+            if (afterLength > 0)
+            {
+                yield return CreatePaleMarker(lineRemoved.Offset + afterBegin, afterLength, GetRemovedBackColor());
+            }
         }
+        else
+        {
+            int length = lineRemoved.Length - lineStartOffset;
+            if (length > 0)
+            {
+                yield return CreatePaleMarker(lineRemoved.Offset + lineStartOffset, length, GetRemovedBackColor());
+            }
+        }
+
+        yield break;
+
+        static TextMarker CreatePaleMarker(int offset, int length, Color color)
+            => CreateTextMarker(offset, length, ColorHelper.DimColor(ColorHelper.DimColor(color)));
+
+        static TextMarker CreateTextMarker(int offset, int length, Color color)
+            => new(offset, length, TextMarkerType.SolidBlock, color, ColorHelper.GetForeColorForBackColor(color));
+
+        static Color GetAddedBackColor() => AppColor.AnsiTerminalGreenBackNormal.GetThemeColor();
+        static Color GetRemovedBackColor() => AppColor.AnsiTerminalRedBackNormal.GetThemeColor();
     }
 
     private void AddExtraPatchHighlighting(IDocument document)
```

Comment:
Maybe rename this function
On 320/372 add comment that the file names in the header is updated
On 323/375 describe the function briefly, similar to the PR description. Maybe in GetDifferenceMarkers instead.
Then you have a starter to understand and improve the algorithm even further